src/Nethermind/Nethermind.Blockchain/Blocks/PotentialCensorshipCache.cs

```diff
@@ -0,0 +1,79 @@
+// SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using Nethermind.Core;
+using Nethermind.Core.Caching;
+using Nethermind.Core.Crypto;
+using Nethermind.Int256;
+using Nethermind.TxPool.Collections;
+
+namespace Nethermind.Blockchain.Blocks;
+
+public class PotentialCensorshipCache
+{
+    private static PotentialCensorshipCache? _instance;
+
+    public static PotentialCensorshipCache Instance()
+    {
+        _instance ??= new PotentialCensorshipCache();
+        return _instance;
+    }
+    private const int CacheSize = 4;
+
+    private readonly LruCache<Hash256, TxCensorshipDetector>
+        _potentialCensorshipCache = new(CacheSize, CacheSize, "potentialCensorshipCache");
+
+    public struct TxCensorshipDetector
+    {
+        public bool PotentialCensorship { get; set; }
+        public ValueHash256 TxHash { get; set; }
+
+        public TxCensorshipDetector(bool potentialCensorship, ValueHash256 txHash)
+        {
+            PotentialCensorship = potentialCensorship;
+            TxHash = txHash;
+        }
+    }
+
+    public PotentialCensorshipCache() { }
+
+    public void Delete(Hash256 blockHash)
+    {
+        _potentialCensorshipCache.Delete(blockHash);
+    }
+
+    public void Cache(ref Block block)
+    {
+        UInt256 maxGasPriceInBlock = 0;
+
+        for (int i = 0; i < block.Transactions.Length; i++)
+        {
+            maxGasPriceInBlock = maxGasPriceInBlock > block.Transactions[i].GasPrice ? maxGasPriceInBlock : block.Transactions[i].GasPrice;
+        }
+
+        TxGasPriceSortedCollection.TxHashGasPricePair pair = TxGasPriceSortedCollection.Instance().GetFirstPair();
+
+        if (pair.GasPrice > maxGasPriceInBlock)
+        {
+            _potentialCensorshipCache.Set(block.Hash, new TxCensorshipDetector(true, pair.Hash));
+        }
+        else
+        {
+            _potentialCensorshipCache.Set(block.Hash, new TxCensorshipDetector(false, pair.Hash));
+        }
+    }
+
+    public bool CensorshipDetected()
+    {
+        ValueHash256 toCompare = _potentialCensorshipCache.GetValues()[0].TxHash;
+        foreach (TxCensorshipDetector suspect in _potentialCensorshipCache.GetValues())
+        {
+            if (!suspect.PotentialCensorship || suspect.TxHash != toCompare)
+            {
+                return false;
+            }
+        }
```

Comment:
I don't get this logic, can you explain?

Reply:
For the cached blocks, if potential censorship is false for any one, there is no censorship. On the other hand, if the highest tx in the TxGasPriceSortedCollection is not the same for all cached blocks, there is no censorship. Reason being, there might be a new tx added to the collection with a higher gas price than the block's maximum after the block was committed. It can happen continuously over a few blocks.